File: frontend/src/api/inventory/reports.ts
import { apiFetch } from '../client';

export type ProjectConsumptionRow = {
  project_id: number;
  item_id: number;
  qty_issued: number;
};

export type ProjectConsumptionQuery = {
  date_from: string;
  date_to: string;
  project_id?: number;
  item_id?: number;
};

export function getProjectConsumption(
  query: ProjectConsumptionQuery,
): Promise<ProjectConsumptionRow[]> {
  const params = new URLSearchParams({
    date_from: query.date_from,
    date_to: query.date_to,
  });

  if (query.project_id !== undefined) params.set('project_id', String(query.project_id));
  if (query.item_id !== undefined) params.set('item_id', String(query.item_id));

  return apiFetch<ProjectConsumptionRow[]>(
    `/api/v1/inventory/reports/project-consumption?${params.toString()}`,
  );
}

File: backend/app/api/v1/inventory/reports.py
from __future__ import annotations

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.core.auth import require_permissions
from app.core.database import get_db
from app.modules.inventory.schemas import (
    ProjectConsumptionQuery,
    ProjectConsumptionRow,
)
from app.modules.inventory.service import get_project_consumption

router = APIRouter(prefix="/reports")


@router.get(
    "/project-consumption",
    response_model=list[ProjectConsumptionRow],
    dependencies=[
        Depends(
            require_permissions(
                {"inventory.reports.project_consumption.read"}
            )
        ),
    ],
)
def api_project_consumption(
    q: ProjectConsumptionQuery = Depends(),
    db: Session = Depends(get_db),
) -> list[ProjectConsumptionRow]:
    rows = get_project_consumption(
        db,
        date_from=q.date_from,
        date_to=q.date_to,
        project_id=q.project_id,
        item_id=q.item_id,
    )
    return [
        ProjectConsumptionRow(
            project_id=project_id,
            item_id=item_id,
            qty_issued=qty_issued,
        )
        for project_id, item_id, qty_issued in rows
    ]
